// @shopify/react-performance/build/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var performance = require('@shopify/performance');
var context = require('./context.js');
var performanceEffect = require('./performance-effect.js');
var PerformanceMark = require('./PerformanceMark.js');
var performanceMark = require('./performance-mark.js');
var LifecycleEventListener = require('./LifecycleEventListener.js');
var lifecycleEventListener = require('./lifecycle-event-listener.js');
var NavigationListener = require('./NavigationListener.js');
var navigationListener = require('./navigation-listener.js');
var PerformanceReport = require('./PerformanceReport.js');
var performanceReport = require('./performance-report.js');
var types = require('./types.js');



exports.PerformanceContext = context.PerformanceContext;
exports.usePerformanceEffect = performanceEffect.usePerformanceEffect;
exports.PerformanceMark = PerformanceMark.PerformanceMark;
exports.usePerformanceMark = performanceMark.usePerformanceMark;
exports.LifecycleEventListener = LifecycleEventListener.LifecycleEventListener;
exports.useLifecycleEventListener = lifecycleEventListener.useLifecycleEventListener;
exports.NavigationListener = NavigationListener.NavigationListener;
exports.useNavigationListener = navigationListener.useNavigationListener;
exports.PerformanceReport = PerformanceReport.PerformanceReport;
exports.usePerformanceReport = performanceReport.usePerformanceReport;
Object.defineProperty(exports, 'Stage', {
	enumerable: true,
	get: function () { return types.Stage; }
});
Object.keys(performance).forEach(function (k) {
	if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return performance[k]; }
	});
});
