// promjs/histogram.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Histogram = void 0;

var _lodash = require("lodash");

var _collector = require("./collector");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function findMinBucketIndex(ary, num) {
  if (num < ary[ary.length - 1]) {
    for (var i = 0; i < ary.length; i += 1) {
      if (num <= ary[i]) {
        return i;
      }
    }
  }

  return undefined;
}

function getInitialValue(buckets) {
  // Make the skeleton to which values will be saved.
  var entries = (0, _lodash.reduce)(buckets, function (result, b) {
    result[b.toString()] = 0;
    return result;
  }, {
    '+Inf': 0
  });
  return {
    entries: entries,
    sum: 0,
    count: 0,
    raw: []
  };
}

var Histogram =
/*#__PURE__*/
function (_Collector) {
  _inherits(Histogram, _Collector);

  function Histogram() {
    var _this;

    var buckets = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

    _classCallCheck(this, Histogram);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Histogram).call(this)); // Sort to get smallest -> largest in order.

    _defineProperty(_assertThisInitialized(_this), "buckets", void 0);

    _this.buckets = buckets.sort(function (a, b) {
      return a > b ? 1 : -1;
    });

    _this.set(getInitialValue(_this.buckets));

    _this.observe = _this.observe.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(Histogram, [{
    key: "observe",
    value: function observe(value, labels) {
      var metric = this.get(labels);

      if (metric == null) {
        // Create a metric for the labels.
        metric = this.set(getInitialValue(this.buckets), labels).get(labels);
      }

      metric.value.raw.push(value);
      metric.value.entries['+Inf'] += 1;
      var minBucketIndex = findMinBucketIndex(this.buckets, value);

      if (minBucketIndex != null) {
        for (var i = minBucketIndex; i < this.buckets.length; i += 1) {
          var val = metric.value.entries[this.buckets[i].toString()];
          metric.value.entries[this.buckets[i].toString()] = val + 1;
        }
      }

      metric.value.sum = (0, _lodash.sum)(metric.value.raw);
      metric.value.count += 1;
      return this;
    }
  }, {
    key: "reset",
    value: function reset(labels) {
      this.set(getInitialValue(this.buckets), labels);
    }
  }]);

  return Histogram;
}(_collector.Collector);

exports.Histogram = Histogram;
