// promjs/registry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Registry = void 0;

var _lodash = require("lodash");

var _counter = require("./counter");

var _gauge = require("./gauge");

var _histogram = require("./histogram");

var _utils = require("./utils");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Registry =
/*#__PURE__*/
function () {
  function Registry() {
    _classCallCheck(this, Registry);

    _defineProperty(this, "data", void 0);

    this.data = {
      counter: {},
      gauge: {},
      histogram: {}
    };
  }

  _createClass(Registry, [{
    key: "validateInput",
    value: function validateInput(type, name, help, buckets) {
      // checks for js runtime
      if (String(name) === '') {
        throw new Error('Metric name cannot be empty');
      }

      if (['counter', 'gauge', 'histogram'].indexOf(type) === -1) {
        throw new Error("Unknown metric type ".concat(type));
      }

      if (typeof help !== 'string' && help != null) {
        throw new Error('help must be string or undefined/null');
      }

      if ((0, _lodash.has)(this.data, "".concat(type, ".").concat(name))) {
        throw new Error("A metric with the name '".concat(name, "' already exists for type '").concat(type, "'"));
      }

      if (!Array.isArray(buckets) && buckets != null) {
        throw new Error('buckets must be array or undefined/null');
      }
    }
  }, {
    key: "create",
    value: function create(type, name) {
      var help = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
      var histogramBuckets = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
      this.validateInput(type, name, help, histogramBuckets);
      var instance;

      if (type === 'counter') {
        instance = new _counter.Counter();
        this.data.counter[name] = {
          help: help,
          instance: instance,
          type: type
        };
      } else if (type === 'gauge') {
        instance = new _gauge.Gauge();
        this.data.gauge[name] = {
          help: help,
          instance: instance,
          type: type
        };
      } else {
        instance = new _histogram.Histogram(histogramBuckets);
        this.data.histogram[name] = {
          help: help,
          instance: instance,
          type: type
        };
      }

      return instance;
    }
    /**
     * Returns a string in the prometheus' desired format
     * More info: https://prometheus.io/docs/concepts/data_model/
     * Loop through each metric type (counter, histogram, etc);
     *
     * @return {string}
     */

  }, {
    key: "metrics",
    value: function metrics() {
      return (0, _lodash.reduce)(this.data, function (output, metrics, type) {
        return output + (0, _lodash.reduce)(metrics, function (src, metric, name) {
          var values = metric.instance.collect();
          var result = src;

          if (metric.help.length > 0) {
            result += "# HELP ".concat(name, " ").concat(metric.help, "\n");
          }

          result += "# TYPE ".concat(name, " ").concat(type, "\n"); // Each metric can have many labels. Iterate over each and append to the string.

          result += (0, _lodash.reduce)(values, function (str, value) {
            var formatted = type === 'histogram' ? (0, _utils.formatHistogramOrSummary)(name, value) : (0, _utils.formatCounterOrGauge)(name, value);
            return str + formatted;
          }, '');
          return result;
        }, '');
      }, '');
    }
  }, {
    key: "reset",
    value: function reset() {
      (0, _lodash.each)(this.data, function (metrics) {
        (0, _lodash.each)(metrics, function (_ref) {
          var instance = _ref.instance;
          instance.resetAll();
        });
      });
      return this;
    }
  }, {
    key: "clear",
    value: function clear() {
      this.data = {
        counter: {},
        gauge: {},
        histogram: {}
      };
      return this;
    }
  }, {
    key: "get",
    value: function get(type, name) {
      var registryItems = type != null ? [this.data[type]] : (0, _lodash.valuesIn)(this.data);
      var metric = registryItems.find(function (v) {
        return (0, _lodash.has)(v, name);
      });
      return metric != null ? metric[name].instance : undefined;
    }
  }]);

  return Registry;
}();

exports.Registry = Registry;
