// @aelesia/commons/dist/src/aws/dynamodb/AwsDynamodb.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsDynamodb = void 0;
var AWS = __importStar(require("aws-sdk"));
var Error_1 = require("../../error/Error");
var Misc_1 = require("../../collections/Misc");
var AwsDynamodb = (function () {
    function AwsDynamodb(region, table) {
        AWS.config.update({ region: region });
        this.ddb = new AWS.DynamoDB.DocumentClient();
        this.table = table;
    }
    AwsDynamodb.prototype.select = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var params;
            var _this = this;
            return __generator(this, function (_a) {
                params = {
                    TableName: this.table,
                    Key: { id: id }
                };
                return [2, new Promise(function (resolve, reject) {
                        _this.ddb.get(params, function (err, data) {
                            if (err) {
                                reject(err);
                            }
                            else if (!data || !data.Item) {
                                reject(new Error("No items retrieved from: " + _this.table + " : {id:" + id + "}"));
                            }
                            else {
                                resolve(Misc_1.Misc.convertISOToDateObj(data.Item));
                            }
                        });
                    })];
            });
        });
    };
    AwsDynamodb.prototype.insert = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var params;
            var _this = this;
            return __generator(this, function (_a) {
                params = {
                    TableName: this.table,
                    Item: Misc_1.Misc.convertDateObjToISO(data)
                };
                return [2, new Promise(function (resolve, reject) {
                        _this.ddb.put(params, function (err, _data) {
                            if (err) {
                                reject(err);
                            }
                            else {
                                resolve(data);
                            }
                        });
                    })];
            });
        });
    };
    AwsDynamodb.prototype.scan = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2, new Promise(function (resolve, reject) {
                        _this.ddb.scan({ TableName: _this.table }, function (err, data) {
                            if (err) {
                                reject(err);
                            }
                            else if (!data || !data.Items || data.Items.length <= 0) {
                                resolve([]);
                            }
                            else {
                                resolve(Misc_1.Misc.convertISOToDateObj(data.Items));
                            }
                        });
                    })];
            });
        });
    };
    AwsDynamodb.prototype.update = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var current_data, new_data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.select(data.id)];
                    case 1:
                        current_data = _a.sent();
                        new_data = __assign(__assign({}, current_data), data);
                        return [4, this.insert(new_data)];
                    case 2:
                        _a.sent();
                        return [2, new_data];
                }
            });
        });
    };
    AwsDynamodb.prototype.delete = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2, new Promise(function (resolve, reject) {
                        _this.ddb.delete({ TableName: _this.table, Key: { id: id } }, function (err, _data) {
                            if (err) {
                                reject(err);
                            }
                            else {
                                resolve();
                            }
                        });
                    })];
            });
        });
    };
    AwsDynamodb.prototype.deleteAll = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                throw new Error_1.NotImplementedErr('AwsDynamodb.deleteAll() has not yet been implemented');
            });
        });
    };
    return AwsDynamodb;
}());
exports.AwsDynamodb = AwsDynamodb;
