// @aelesia/commons/dist/src/aws/dynamodb/FileSystemDatabase.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSystemDatabase = void 0;
var fs = __importStar(require("fs"));
var Regex_1 = require("../../collections/Regex");
var File = (function () {
    function File(name, directory) {
        if (directory === void 0) { directory = './db'; }
        this.path = directory + "/" + name + ".json";
        if (!fs.existsSync(this.path)) {
            if (!fs.existsSync(directory)) {
                fs.mkdirSync(directory);
            }
            fs.appendFileSync(this.path, JSON.stringify({}));
        }
    }
    File.prototype.read = function () {
        var file = fs.readFileSync(this.path);
        return JSON.parse(file.toString(), function (key, value) {
            if (typeof value == 'string' && Regex_1.Regex.is.isoDate(value)) {
                return new Date(value);
            }
            return value;
        });
    };
    File.prototype.write = function (object) {
        fs.writeFileSync(this.path, JSON.stringify(object, null, 2));
    };
    return File;
}());
var FileSystemDatabase = (function () {
    function FileSystemDatabase(table, folderDir) {
        this.table = table;
        this.db = new File(table, folderDir);
    }
    FileSystemDatabase.prototype.insert = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                result = this.db.read();
                result[data.id] = data;
                this.db.write(result);
                return [2, data];
            });
        });
    };
    FileSystemDatabase.prototype.scan = function () {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                result = this.db.read();
                return [2, Object.keys(result).map(function (id) { return result[id]; })];
            });
        });
    };
    FileSystemDatabase.prototype.select = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var value;
            return __generator(this, function (_a) {
                value = this.db.read()[id];
                if (value) {
                    return [2, value];
                }
                throw Error("[DB] [" + this.table + "] No such value: " + id);
            });
        });
    };
    FileSystemDatabase.prototype.update = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var current_data, new_data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.select(data.id)];
                    case 1:
                        current_data = _a.sent();
                        new_data = __assign(__assign({}, current_data), data);
                        return [4, this.insert(new_data)];
                    case 2:
                        _a.sent();
                        return [2, new_data];
                }
            });
        });
    };
    FileSystemDatabase.prototype.delete = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                result = this.db.read();
                delete result[id];
                this.db.write(result);
                return [2];
            });
        });
    };
    FileSystemDatabase.prototype.deleteAll = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.db.write({});
                return [2];
            });
        });
    };
    return FileSystemDatabase;
}());
exports.FileSystemDatabase = FileSystemDatabase;
