// @aelesia/commons/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.env = exports.Throw = exports.sleep = exports.loop = exports.Err = exports.Misc = exports.Regex = exports.Format = exports.Obj = exports.Time = exports.Rand = exports.Str = exports.Arr = exports.Date_ = exports._ = void 0;
var DateUtil_1 = require("./src/collections/util/DateUtil");
Object.defineProperty(exports, "Date_", { enumerable: true, get: function () { return DateUtil_1.DateUtil; } });
var StringUtil_1 = require("./src/collections/util/StringUtil");
Object.defineProperty(exports, "Str", { enumerable: true, get: function () { return StringUtil_1.StringUtil; } });
var ArrayUtil_1 = require("./src/collections/util/ArrayUtil");
Object.defineProperty(exports, "Arr", { enumerable: true, get: function () { return ArrayUtil_1.ArrayUtil; } });
var RandomUtil_1 = require("./src/collections/util/RandomUtil");
Object.defineProperty(exports, "Rand", { enumerable: true, get: function () { return RandomUtil_1.RandomUtil; } });
var TimeUtil_1 = require("./src/collections/util/TimeUtil");
Object.defineProperty(exports, "Time", { enumerable: true, get: function () { return TimeUtil_1.TimeUtil; } });
var Format_1 = require("./src/collections/Format");
Object.defineProperty(exports, "Format", { enumerable: true, get: function () { return Format_1.Format; } });
var Regex_1 = require("./src/collections/Regex");
Object.defineProperty(exports, "Regex", { enumerable: true, get: function () { return Regex_1.Regex; } });
var Misc_1 = require("./src/collections/Misc");
Object.defineProperty(exports, "Misc", { enumerable: true, get: function () { return Misc_1.Misc; } });
var Control_1 = require("./src/general/Control");
Object.defineProperty(exports, "loop", { enumerable: true, get: function () { return Control_1.loop; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return Control_1.sleep; } });
Object.defineProperty(exports, "Throw", { enumerable: true, get: function () { return Control_1.Throw; } });
Object.defineProperty(exports, "env", { enumerable: true, get: function () { return Control_1.env; } });
var Err = __importStar(require("./src/error/Error"));
exports.Err = Err;
var ObjectUtil_1 = require("./src/collections/util/ObjectUtil");
Object.defineProperty(exports, "Obj", { enumerable: true, get: function () { return ObjectUtil_1.ObjectUtil; } });
exports._ = {
    date: DateUtil_1.DateUtil,
    arr: ArrayUtil_1.ArrayUtil,
    str: StringUtil_1.StringUtil,
    rand: RandomUtil_1.RandomUtil,
    obj: ObjectUtil_1.ObjectUtil,
    time: TimeUtil_1.TimeUtil,
    format: Format_1.Format,
    regex: Regex_1.Regex
};
var FakerFactory_1 = require("./src/dummy/FakerFactory");
Object.defineProperty(exports, "FakerFactory", { enumerable: true, get: function () { return FakerFactory_1.FakerFactory; } });
var Patch_1 = require("./src/dummy/Patch");
Object.defineProperty(exports, "Patch", { enumerable: true, get: function () { return Patch_1.Patch; } });
