// @shopify/performance/build/cjs/inflight.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utilities = require('./utilities.js');
var types = require('./types.js');
var navigation = require('./navigation.js');

function replaceResourceEvent(event, oldEvent) {
  if (event.type !== types.EventType.ScriptDownload && event.type !== types.EventType.StyleDownload || oldEvent.type !== types.EventType.ScriptDownload && oldEvent.type !== types.EventType.StyleDownload) {
    return false;
  }

  return event.type === oldEvent.type && event.metadata.name === oldEvent.metadata.name;
}

const MATCHES_CHECK_MAP = new Map([// In Safari, we occasionally get the same "resource" event twice.
// This check looks for a resource event with the same name, and replaces
// it if found.
[types.EventType.ScriptDownload, replaceResourceEvent], [types.EventType.StyleDownload, replaceResourceEvent]]);

function defaultEqualityCheck({
  type
}, {
  type: otherType
}) {
  return type === otherType;
}

class InflightNavigation {
  constructor({
    timeOrigin,
    start = utilities.now(),
    target = window.location.pathname
  }, metadata) {
    this.metadata = metadata;
    this.timeOrigin = void 0;
    this.start = void 0;
    this.target = void 0;
    this.events = [];
    this.timeOrigin = timeOrigin;
    this.start = this.normalize(start);
    this.target = target;
  }

  event(event, // Users can either force "matching" events to be replaced, or provide
  // a function that checks for an event that should be replaced. This may
  // be necessary if events can be triggered multiple times because of
  // browser inconsistencies.
  replaceExisting = false) {
    const normalizedEvent = { ...event,
      start: this.normalize(event.start)
    };

    if (replaceExisting) {
      const check = typeof replaceExisting === 'function' ? replaceExisting : MATCHES_CHECK_MAP.get(event.type) || defaultEqualityCheck;
      const existingIndex = this.events.findIndex(oldEvent => check(event, oldEvent));

      if (existingIndex >= 0) {
        this.events.splice(existingIndex, 1, normalizedEvent);
      } else {
        this.events.push(normalizedEvent);
      }
    } else {
      this.events.push(normalizedEvent);
    }
  }

  cancel(timeStamp = utilities.now()) {
    return this.end(timeStamp, types.NavigationResult.Cancelled);
  }

  timeout(timeStamp = utilities.now()) {
    return this.end(timeStamp, types.NavigationResult.TimedOut);
  }

  finish(timeStamp = utilities.now()) {
    return this.end(timeStamp, types.NavigationResult.Finished);
  }

  end(timeStamp, result) {
    return new navigation.Navigation({
      target: this.target,
      start: this.start,
      duration: this.normalize(timeStamp) - this.start,
      events: this.events.sort((eventOne, eventTwo) => eventOne.start - eventTwo.start),
      result
    }, this.metadata);
  }

  normalize(timeStamp) {
    return this.timeOrigin + timeStamp;
  }

}

exports.InflightNavigation = InflightNavigation;
